bentoml/_internal/frameworks/common/pytorch.py

```diff
@@ -0,0 +1,183 @@
+        name: t.Optional[str] = None,
+    ):
+        super().__init__(tag=tag, name=name)
+
+        self._predict_fn_name = predict_fn_name
+        self._partial_kwargs = partial_kwargs or dict()
+
+        self._predict_fn: t.Callable[..., torch.Tensor]
+        self._no_grad_context: t.Optional[contextlib.ExitStack] = None
+        self._model: t.Optional["ModelType"] = None
+
+    @property
+    def _device_id(self):
+        if self._on_gpu:
+            return "cuda"
+        else:
+            return "cpu"
+
+    @property
+    def _num_threads(self) -> int:
+        if self._on_gpu:
+            return 1
+        return int(round(self.resource_quota.cpu))
+
+    @property
+    def num_replica(self) -> int:
+        if self._on_gpu:
+            return torch.cuda.device_count()
+        return 1
+
+    def _configure(self) -> None:
+        if self._on_gpu:
+            torch.set_default_tensor_type("torch.cuda.FloatTensor")
+        else:
+            torch.set_num_threads(self._num_threads)
+            torch.set_default_tensor_type("torch.FloatTensor")
+
+    @property
+    def _on_gpu(self) -> bool:
+        if self.resource_quota.on_gpu:
+            if torch.cuda.is_available():
+                return True
+            else:
+                logger.warning(
+                    "GPU is not available, but GPU resource is requested. "
+                    "Falling back to CPU."
+                )
+        return False
+
+    @abstractmethod
+    def _load_model(self):
+        raise NotImplementedError
+
+    def _setup(self) -> None:
+        self._no_grad_context = contextlib.ExitStack()
+        self._no_grad_context.enter_context(torch.no_grad())
+        if get_pkg_version("torch").startswith("1.9"):
+            # inference mode is required for PyTorch version 1.9.*
+            self._no_grad_context.enter_context(torch.inference_mode())
+
+        self._configure()
+        model = self._load_model()
+        model.eval()
+        if self._on_gpu:
+            self._model = parallel.DataParallel(model)
+            torch.cuda.empty_cache()
+        else:
+            self._model = model
+        raw_predict_fn = getattr(self._model, self._predict_fn_name)
+        self._predict_fn = functools.partial(raw_predict_fn, **self._partial_kwargs)
+
+    def _shutdown(self) -> None:
+        if self._no_grad_context is not None:
+            self._no_grad_context.close()
+            self._no_grad_context = None
+
+    def _run_batch(
+        self,
+        *args: t.Union["ext.NpNDArray", torch.Tensor],
+        **kwargs: t.Union["ext.NpNDArray", torch.Tensor],
+    ) -> torch.Tensor:
+
+        params = Params[t.Union["ext.NpNDArray", torch.Tensor]](*args, **kwargs)
+
+        def _mapping(item: t.Union["ext.NpNDArray", torch.Tensor]) -> torch.Tensor:
+            if LazyType["ext.NpNDArray"]("numpy.ndarray").isinstance(item):
+                item = torch.Tensor(item, device=self._device_id)
+            return item
```

Comment:
If it's already a tensor, we need to call `to(device)` on it as well, right?

Reply:
init with `self._device_id` so no need to call `to(device)`

Reply:
@aarnphm @bojiang  I think you mean in `self._configure()` we have set default tensor storage type? But if the tensor is deserialized from DataContainer payload which is serialized in another machine and these 2 machine have mismatched device, then the tensor won't be converted to the proper device automatically. Also in my simple benchmark `to(device)` won't cost a lot if `from_device` and `to_device` is the same

Here's my simple demo:

https://gist.github.com/larme/7640f2d97d45c92946f1573b3eaaa0be

Reply:
I think calling `to` shouldn't affect perf that much. I think we just need to make sure that device is consistent across, so that https://github.com/bentoml/BentoML/issues/2230 won't happen.

Reply:
I think this change will make device more consistent?

Reply:
sounds good then